File: implementations/java-springboot/examples/src/main/java/com/example/accountservice/notification/infrastructure/persistence/SentEmail.java
package com.example.accountservice.notification.infrastructure.persistence;

import com.example.accountservice.common.IdGenerator;
import jakarta.persistence.*;

import java.time.LocalDateTime;

@Entity
@Table(name = "sent_email")
public class SentEmail {

    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    private Long id;

    @Column(nullable = false, unique = true)
    private String sentEmailId;

    @Column(nullable = false)
    private String accountId;

    @Column(nullable = false)
    private String eventType;

    @Column(nullable = false)
    private String recipient;

    @Column(nullable = false)
    private String subject;

    @Column(nullable = false)
    private String sesMessageId;

    @Column(nullable = false)
    private LocalDateTime sentAt;

    protected SentEmail() {}

    public static SentEmail create(String accountId, String eventType, String recipient, String subject, String sesMessageId) {
        SentEmail sentEmail = new SentEmail();
        sentEmail.sentEmailId = IdGenerator.generate();
        sentEmail.accountId = accountId;
        sentEmail.eventType = eventType;
        sentEmail.recipient = recipient;
        sentEmail.subject = subject;
        sentEmail.sesMessageId = sesMessageId;
        sentEmail.sentAt = LocalDateTime.now();
        return sentEmail;
    }

    public String getSentEmailId() { return sentEmailId; }
    public String getAccountId() { return accountId; }
    public String getEventType() { return eventType; }
    public String getRecipient() { return recipient; }
    public String getSubject() { return subject; }
    public String getSesMessageId() { return sesMessageId; }
    public LocalDateTime getSentAt() { return sentAt; }
}

File: implementations/kotlin-springboot/examples/src/main/kotlin/com/example/accountservice/account/infrastructure/notification/persistence/SentEmail.kt
package com.example.accountservice.account.infrastructure.notification.persistence

import com.example.accountservice.common.generateId
import jakarta.persistence.Column
import jakarta.persistence.Entity
import jakarta.persistence.GeneratedValue
import jakarta.persistence.GenerationType
import jakarta.persistence.Id
import jakarta.persistence.Table
import java.time.LocalDateTime

@Entity
@Table(name = "sent_emails")
class SentEmail protected constructor() {

    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    var id: Long? = null
        private set

    @Column(nullable = false, unique = true)
    var sentEmailId: String = ""
        private set

    @Column(nullable = false)
    var accountId: String = ""
        private set

    @Column(nullable = false)
    var eventType: String = ""
        private set

    @Column(nullable = false)
    var recipient: String = ""
        private set

    @Column(nullable = false)
    var subject: String = ""
        private set

    @Column(nullable = false)
    var sesMessageId: String = ""
        private set

    @Column(nullable = false)
    var sentAt: LocalDateTime = LocalDateTime.now()
        private set

    companion object {
        fun create(
            accountId: String,
            eventType: String,
            recipient: String,
            subject: String,
            sesMessageId: String,
        ): SentEmail =
            SentEmail().apply {
                this.sentEmailId = generateId()
                this.accountId = accountId
                this.eventType = eventType
                this.recipient = recipient
                this.subject = subject
                this.sesMessageId = sesMessageId
                this.sentAt = LocalDateTime.now()
            }
    }
}
